File: typescript/src/auth/native.ts
import { Keypair } from "@solana/web3.js";
import bs58 from "bs58";
import nacl from "tweetnacl";
import { generateSigninMessage } from "./index";

export interface SignedLogin {
  message: string;
  signature_bs58: string;
  pubkey_bytes: Uint8Array;
}

export function signLoginMessage(keypair: Keypair, nonce: string): SignedLogin {
  const messageBytes = generateSigninMessage(nonce);
  const signature = nacl.sign.detached(messageBytes, keypair.secretKey);

  return {
    message: new TextDecoder().decode(messageBytes),
    signature_bs58: bs58.encode(signature),
    pubkey_bytes: keypair.publicKey.toBytes(),
  };
}

File: rust/src/auth/native.rs
//! Native auth — keypair-based signing.
//!
//! Only available with the `native-auth` feature.

use solana_keypair::Keypair;
use solana_signer::Signer;

use crate::auth::generate_signin_message;

/// Signed login material ready to pass to `client.auth().login_with_message()`.
pub struct SignedLogin {
    pub message: String,
    pub signature_bs58: String,
    pub pubkey_bytes: [u8; 32],
}

/// Sign a login message with a local keypair.
///
/// The `nonce` must be obtained from `client.auth().get_nonce()` first.
/// Returns a [`SignedLogin`] that can be passed directly to
/// `client.auth().login_with_message()`.
pub fn sign_login_message(keypair: &Keypair, nonce: &str) -> SignedLogin {
    let message_bytes = generate_signin_message(nonce);
    let signature = keypair.sign_message(&message_bytes);
    let message = String::from_utf8(message_bytes)
        .expect("generate_signin_message always produces valid UTF-8");

    SignedLogin {
        message,
        signature_bs58: signature.to_string(),
        pubkey_bytes: keypair.pubkey().to_bytes(),
    }
}
